provider/pkg/provider/provider_parameterize.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2025, Pulumi Corporation.
+
+package provider
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/pulumi/pulumi-azure-native/v2/provider/pkg/resources"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// parameterizedPackageMap maps parameterized package names to their schema, in form of serialized JSON. Will be empty
+// if Parameterize() hasn't been called.
+type parameterizedPackageMap struct {
+	sync.Mutex
+	schemas map[string][]byte
```

Comment:
Might be nicer to replace the lock based approach with a `sync.map` - with your methods just adding the typing.
```suggestion
	schemas sync.Map
```